components/board/genericlinux/data.go

```diff
@@ -144,53 +144,40 @@ func getPwmChipDefs(pinDefs []PinDefinition) (map[string]pwmChipData, error) {
 
 	// Now, look for all chips whose names we found.
 	pwmChipsInfo := map[string]pwmChipData{}
+	var found bool
+	sysPrefix := "/sys/class/pwm"
+	files, err := os.ReadDir(sysPrefix)
+
 	for chipName := range pwmChipNames {
-		found := false
-
-		// The boards we support might store their PWM devices in several different locations
-		// within /sys/devices. If it turns out that every board stores its PWM chips in a
-		// different location, we should rethink this. Maybe a breadth-first search over all of
-		// /sys/devices/? but for now, this is good enough.
-		directoriesToSearch := []string{
-			// Example boards which use each location:
-			// Jetson Orin AGX       BeagleBone AI64                     Intel UP 4000
-			"/sys/devices/platform", "/sys/devices/platform/bus@100000", "/sys/devices/pci0000:00",
+		found = false
+		if err != nil {
+			return nil, err
 		}
-		for _, baseDir := range directoriesToSearch {
-			// For exactly one baseDir, there should be a directory at <baseDir>/<chipName>/pwm/,
-			// which contains a single sub-directory whose name is mirrored in /sys/class/pwm.
-			// That's the one we want to use.
-			chipDir := fmt.Sprintf("%s/%s/pwm", baseDir, chipName)
-			files, err := os.ReadDir(chipDir)
+
+		for _, file := range files {
+			if !strings.HasPrefix(file.Name(), "pwmchip") {
+				continue
+			}
+
+			// look at symlinks to find the correct chip
+			symlink, err := os.Readlink(filepath.Join(sysPrefix, file.Name()))
 			if err != nil {
-				continue // This was the wrong directory; try the next baseDir.
+				return nil, err
 			}
 
-			// We've found what looks like the right place to look for things! Now, find the name
-			// of the chip that should be mirrored in /sys/class/pwm/.
-			for _, file := range files {
-				if !(strings.Contains(file.Name(), "pwmchip") && file.IsDir()) {
-					continue
-				}
+			var chipPath string
+			if strings.Contains(symlink, chipName) {
 				found = true
-				chipPath := fmt.Sprintf("/sys/class/pwm/%s", file.Name())
-
+				chipPath = fmt.Sprintf("/sys/class/pwm/%s", file.Name())
```

Comment:
I like the way you gave the initial path a name higher up! Let's use it here.
```suggestion
				chipPath = fmt.Sprintf("%s/%s", sysPrefix, file.Name())
```

...and with that insight, I suspect we should use [`filepath.Join`](https://pkg.go.dev/path/filepath#Join) instead. I wish I'd thought of that months ago when I wrote the original version. :confounded: